File: agent/course_agent.py
import asyncio
import json
import os
from collections.abc import Awaitable, Callable
from typing import TypeVar
from uuid import uuid4

from google.adk.agents import Agent
from google.adk.runners import Runner
from google.adk.sessions import InMemorySessionService
from google.genai import types

from models.schemas import AssignmentAnalysis

from .prompts import (
    SYSTEM_INSTRUCTION,
    build_assignment_prompt,
    build_pdf_assignment_prompt,
)

APP_NAME = "courseflow"
MODEL_NAME = "gemini-3.5-flash"
MAX_PROVIDER_ATTEMPTS = 3
TRANSIENT_PROVIDER_STATUS_CODES = {429, 500, 502, 503, 504}
T = TypeVar("T")

assignment_agent = Agent(
    name="course_assignment_agent",
    model=MODEL_NAME,
    instruction=SYSTEM_INSTRUCTION,
    output_schema=AssignmentAnalysis,
    generate_content_config=types.GenerateContentConfig(temperature=0.2),
)

def _is_transient_provider_error(error: Exception) -> bool:
    visited: set[int] = set()
    current: BaseException | None = error
    while current is not None and id(current) not in visited:
        visited.add(id(current))
        for attribute in ("status_code", "code"):
            status = getattr(current, attribute, None)
            try:
                if int(status) in TRANSIENT_PROVIDER_STATUS_CODES:
                    return True
            except (TypeError, ValueError):
                pass
        nested = getattr(current, "error", None)
        current = (
            nested
            if isinstance(nested, BaseException)
            else current.__cause__ or current.__context__
        )
    return False


async def _run_with_provider_retries(
    operation: Callable[[], Awaitable[T]],
    *,
    delay_seconds: float = 0.25,
) -> T:
    for attempt in range(MAX_PROVIDER_ATTEMPTS):
        try:
            return await operation()
        except Exception as error:
            if (
                attempt == MAX_PROVIDER_ATTEMPTS - 1
                or not _is_transient_provider_error(error)
            ):
                raise
            await asyncio.sleep(delay_seconds * (2**attempt))
    raise RuntimeError("Provider retry loop ended unexpectedly.")


async def _run_adk_once(prompt: str) -> str:
    user_id = "courseflow-api"
    session_id = uuid4().hex
    session_service = InMemorySessionService()
    runner = Runner(
        agent=assignment_agent,
        app_name=APP_NAME,
        session_service=session_service,
    )

    await session_service.create_session(
        app_name=APP_NAME,
        user_id=user_id,
        session_id=session_id,
    )

    events = runner.run_async(
        user_id=user_id,
        session_id=session_id,
        new_message=types.Content(
            role="user",
            parts=[types.Part(text=prompt)],
        ),
    )

    response_text = None
    async for event in events:
        if not event.is_final_response() or not event.content:
            continue
        response_text = next(
            (
                part.text
                for part in event.content.parts or []
                if getattr(part, "text", None)
            ),
            None,
        )

    if not response_text:
        raise RuntimeError("The ADK agent returned no final structured response.")
    return response_text


async def _run_agent_async(prompt: str) -> AssignmentAnalysis:
    response_text = await _run_with_provider_retries(
        lambda: _run_adk_once(prompt)
    )

    try:
        response_data = json.loads(response_text)
    except json.JSONDecodeError as error:
        raise RuntimeError("The ADK agent returned invalid JSON.") from error

    return AssignmentAnalysis.model_validate(response_data)


def analyze_assignment(assignment_text: str) -> AssignmentAnalysis:
    """Analyze one assignment through the Google ADK async runner."""
    return asyncio.run(
        _run_agent_async(build_assignment_prompt(assignment_text))
    )


def analyze_assignment_pages(page_blocks: str) -> AssignmentAnalysis:
    """Analyze page-marked PDF text through the same Google ADK agent."""
    return asyncio.run(_run_agent_async(build_pdf_assignment_prompt(page_blocks)))


class SourceReferenceError(ValueError):
    """Raised when the model cites a file or page outside the upload."""


def validate_source_references(
    analysis: AssignmentAnalysis,
    available_pages: dict[str, set[int]],
) -> None:
    evidence_items = list(analysis.deadline_evidence)
    evidence_items.extend(
        evidence
        for sourced_fact in analysis.deliverable_evidence
        for evidence in sourced_fact.evidence
    )
    evidence_items.extend(
        evidence
        for sourced_fact in analysis.requirement_evidence
        for evidence in sourced_fact.evidence
    )
    evidence_items.extend(
        evidence for task in analysis.tasks for evidence in task.evidence
    )

    for evidence in evidence_items:
        if evidence.source_file not in available_pages:
            raise SourceReferenceError(
                f"Gemini cited an uploaded file that was not supplied: "
                f"{evidence.source_file}"
            )
        if evidence.page_number is None:
            raise SourceReferenceError(
                f"Gemini did not provide a page number for {evidence.source_file}."
            )
        if evidence.page_number not in available_pages[evidence.source_file]:
            raise SourceReferenceError(
                f"Gemini cited page {evidence.page_number} of "
                f"{evidence.source_file}, but that page does not exist."
            )

    if analysis.deadline is not None and not analysis.deadline_evidence:
        raise SourceReferenceError(
            "The extracted deadline did not include source evidence."
        )
    deliverable_evidence = {
        sourced_fact.fact: sourced_fact.evidence
        for sourced_fact in analysis.deliverable_evidence
    }
    for deliverable in analysis.deliverables:
        if not deliverable_evidence.get(deliverable):
            raise SourceReferenceError(
                f"The deliverable {deliverable!r} did not include source evidence."
            )
    requirement_evidence = {
        sourced_fact.fact: sourced_fact.evidence
        for sourced_fact in analysis.requirement_evidence
    }
    for requirement in analysis.requirements:
        if not requirement_evidence.get(requirement):
            raise SourceReferenceError(
                f"The requirement {requirement!r} did not include source evidence."
            )
    for task in analysis.tasks:
        if task.source_requirement and not task.evidence:
            raise SourceReferenceError(
                f"The task {task.title!r} did not include source evidence."
            )

File: src/agent/courseAgent.ts
import { GoogleGenAI, Type } from '@google/genai';
import { AssignmentAnalysis, SourceEvidence } from '../types.js';
import {
  SYSTEM_INSTRUCTION,
  buildAssignmentPrompt,
  buildPdfAssignmentPrompt,
} from './prompts.js';

export class SourceReferenceError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'SourceReferenceError';
  }
}

const sourceEvidenceSchema = {
  type: Type.OBJECT,
  properties: {
    source_file: { type: Type.STRING },
    page_number: { type: Type.INTEGER },
    source_snippet: { type: Type.STRING },
  },
  required: ['source_file', 'page_number', 'source_snippet'],
};

const evidenceBackedFactSchema = {
  type: Type.OBJECT,
  properties: {
    fact: { type: Type.STRING },
    is_optional: { type: Type.BOOLEAN, nullable: true },
    evidence: {
      type: Type.ARRAY,
      items: sourceEvidenceSchema,
    },
  },
  required: ['fact', 'evidence'],
};

const taskEstimateSchema = {
  type: Type.OBJECT,
  properties: {
    task_id: { type: Type.STRING, nullable: true },
    title: { type: Type.STRING },
    description: { type: Type.STRING },
    source_requirement: { type: Type.STRING },
    dependencies: {
      type: Type.ARRAY,
      items: { type: Type.STRING },
    },
    optimistic_minutes: { type: Type.INTEGER },
    expected_minutes: { type: Type.INTEGER },
    pessimistic_minutes: { type: Type.INTEGER },
    confidence: {
      type: Type.STRING,
      enum: ['low', 'medium', 'high'],
    },
    estimation_reason: { type: Type.STRING },
    is_optional: {
      type: Type.BOOLEAN,
      description: 'True if this task is optional or bonus work; false if mandatory.',
    },
    evidence: {
      type: Type.ARRAY,
      items: sourceEvidenceSchema,
    },
  },
  required: [
    'title',
    'description',
    'source_requirement',
    'dependencies',
    'optimistic_minutes',
    'expected_minutes',
    'pessimistic_minutes',
    'confidence',
    'estimation_reason',
    'is_optional',
    'evidence',
  ],
};

export const assignmentAnalysisSchema = {
  type: Type.OBJECT,
  properties: {
    title: { type: Type.STRING },
    deadline: { type: Type.STRING, nullable: true },
    deadline_iso: { type: Type.STRING, nullable: true },
    deadline_evidence: {
      type: Type.ARRAY,
      items: sourceEvidenceSchema,
    },
    deliverables: {
      type: Type.ARRAY,
      items: { type: Type.STRING },
    },
    deliverable_evidence: {
      type: Type.ARRAY,
      items: evidenceBackedFactSchema,
    },
    requirements: {
      type: Type.ARRAY,
      items: { type: Type.STRING },
    },
    requirement_evidence: {
      type: Type.ARRAY,
      items: evidenceBackedFactSchema,
    },
    ambiguities: {
      type: Type.ARRAY,
      items: { type: Type.STRING },
    },
    tasks: {
      type: Type.ARRAY,
      items: taskEstimateSchema,
    },
  },
  required: [
    'title',
    'deadline_evidence',
    'deliverables',
    'deliverable_evidence',
    'requirements',
    'requirement_evidence',
    'ambiguities',
    'tasks',
  ],
};

function getGenAI(): GoogleGenAI {
  const apiKey =
    process.env.GEMINI_API_KEY || process.env.GOOGLE_API_KEY;
  if (!apiKey) {
    throw new Error('GEMINI_API_KEY is not configured.');
  }
  return new GoogleGenAI({
    apiKey,
    httpOptions: {
      headers: {
        'User-Agent': 'aistudio-build',
      },
    },
  });
}

const MODELS = ['gemini-3.7-flash', 'gemini-3.5-flash', 'gemini-3.5-flash-lite'];

export function validateEvidenceCompleteness(
  analysis: AssignmentAnalysis
): void {
  if (
    analysis.deadline !== null &&
    analysis.deadline !== undefined &&
    (!analysis.deadline_evidence || analysis.deadline_evidence.length === 0)
  ) {
    throw new SourceReferenceError(
      'The extracted deadline did not include source evidence.'
    );
  }

  const deliverableEvidence = new Map<string, SourceEvidence[]>();
  for (const sourcedFact of analysis.deliverable_evidence || []) {
    deliverableEvidence.set(sourcedFact.fact, sourcedFact.evidence || []);
  }

  for (const deliverable of analysis.deliverables || []) {
    const evidence = deliverableEvidence.get(deliverable);
    if (!evidence || evidence.length === 0) {
      throw new SourceReferenceError(
        `The deliverable ${JSON.stringify(deliverable)} did not include source evidence.`
      );
    }
  }

  const requirementEvidence = new Map<string, SourceEvidence[]>();
  for (const sourcedFact of analysis.requirement_evidence || []) {
    requirementEvidence.set(sourcedFact.fact, sourcedFact.evidence || []);
  }

  for (const requirement of analysis.requirements || []) {
    const evidence = requirementEvidence.get(requirement);
    if (!evidence || evidence.length === 0) {
      throw new SourceReferenceError(
        `The requirement ${JSON.stringify(requirement)} did not include source evidence.`
      );
    }
  }

  for (const task of analysis.tasks || []) {
    if (task.source_requirement && (!task.evidence || task.evidence.length === 0)) {
      throw new SourceReferenceError(
        `The task ${JSON.stringify(task.title)} did not include source evidence.`
      );
    }
  }
}

export function parseGeminiAnalysis(
  text: string,
  requireSourceEvidence: boolean
): AssignmentAnalysis {
  const parsed: AssignmentAnalysis = JSON.parse(text);

  if (requireSourceEvidence) {
    validateEvidenceCompleteness(parsed);
  }

  // Validate estimate order: optimistic <= expected <= pessimistic
  if (parsed.tasks) {
    for (const task of parsed.tasks) {
      if (
        !(
          task.optimistic_minutes <= task.expected_minutes &&
          task.expected_minutes <= task.pessimistic_minutes
        )
      ) {
        const sorted = [
          task.optimistic_minutes,
          task.expected_minutes,
          task.pessimistic_minutes,
        ].sort((a, b) => a - b);
        task.optimistic_minutes = sorted[0];
        task.expected_minutes = sorted[1];
        task.pessimistic_minutes = sorted[2];
      }
    }
  }

  return parsed;
}

async function runGeminiWithRetry(
  prompt: string,
  requireSourceEvidence: boolean
): Promise<AssignmentAnalysis> {
  const ai = getGenAI();
  let lastError: any = null;

  for (const model of MODELS) {
    try {
      const response = await ai.models.generateContent({
        model,
        contents: prompt,
        config: {
          systemInstruction: SYSTEM_INSTRUCTION,
          temperature: 0.2,
          responseMimeType: 'application/json',
          responseSchema: assignmentAnalysisSchema,
        },
      });

      const text = response.text?.trim();
      if (!text) {
        throw new Error('Gemini returned an empty response.');
      }

      const parsed = parseGeminiAnalysis(text, requireSourceEvidence);

      console.log(`[courseAgent] Assignment analysis succeeded using model: ${model}`);
      return parsed;
    } catch (err: any) {
      lastError = err;
      const isUnavailable =
        err?.status === 503 ||
        err?.code === 503 ||
        (typeof err?.message === 'string' &&
          (err.message.includes('503') ||
            err.message.includes('UNAVAILABLE') ||
            err.message.includes('high demand') ||
            err.message.includes('ResourceExhausted')));

      if (isUnavailable) {
        console.log(`[courseAgent] ${model} unavailable (high demand), switching to fallback model...`);
      } else {
        console.log(`[courseAgent] ${model} attempt failed: ${err?.message || err}, switching to fallback...`);
      }
    }
  }

  throw lastError || new Error('Failed to analyze assignment with Gemini.');
}

export async function analyzeAssignment(
  assignmentText: string
): Promise<AssignmentAnalysis> {
  return runGeminiWithRetry(buildAssignmentPrompt(assignmentText), false);
}

export async function analyzeAssignmentPages(
  pageBlocks: string
): Promise<AssignmentAnalysis> {
  return runGeminiWithRetry(buildPdfAssignmentPrompt(pageBlocks), true);
}

export function validateSourceReferences(
  analysis: AssignmentAnalysis,
  availablePages: Map<string, Set<number>>
): void {
  const evidenceItems: SourceEvidence[] = [
    ...(analysis.deadline_evidence || []),
    ...(analysis.deliverable_evidence || []).flatMap((d) => d.evidence || []),
    ...(analysis.requirement_evidence || []).flatMap((r) => r.evidence || []),
    ...(analysis.tasks || []).flatMap((t) => t.evidence || []),
  ];

  for (const evidence of evidenceItems) {
    if (!availablePages.has(evidence.source_file)) {
      throw new SourceReferenceError(
        `Gemini cited an uploaded file that was not supplied: ${evidence.source_file}`
      );
    }
    if (evidence.page_number === undefined || evidence.page_number === null) {
      throw new SourceReferenceError(
        `Gemini did not provide a page number for ${evidence.source_file}.`
      );
    }
    const pagesForFile = availablePages.get(evidence.source_file)!;
    if (!pagesForFile.has(evidence.page_number)) {
      throw new SourceReferenceError(
        `Gemini cited page ${evidence.page_number} of ${evidence.source_file}, but that page does not exist.`
      );
    }
  }

  validateEvidenceCompleteness(analysis);
}
